Feat: Monitoring AI metrics reference (#592)
Kong/developer.konghq.com: app/ai-gateway/monitor-ai-llm-metrics.md

---
title: "Monitor AI LLM metrics"
content_type: reference
layout: reference

products:
  - gateway

tags:
  - ai-gateway
  - monitoring

plugins:
  - prometheus
  - ai-proxy
  - ai-proxy-advanced

min_version:
  gateway: '3.7'

description: "This guide walks you through collecting AI metrics and sending them to Prometheus."

related_resources:
  - text: Kong AI Gateway
    url: /ai-gateway/
  - text: Kong AI Gateway plugins
    url: /plugins/?category=ai
  - text: Status API
    url: /api/gateway/status/v1/

>   - text: Admin API
>     url: /api/gateway/admin-ee/3.9/

Reviewer: ```suggestion
  - text: Admin API
    url: /api/gateway/admin-ee/
  - text: Prometheus plugin
    url: /plugins/prometheus/
```

Author: I didn't add Prometheus because it's already in the list of plugins

---

Kong AI Gateway calls LLM-based services according to the settings of the [AI Proxy](/plugins/ai-proxy/) and [AI Proxy Advanced](/plugins/ai-proxy-advanced/) plugins.
You can aggregate the LLM provider responses to count the number of tokens used by the AI plugins.
If you have defined input and output costs in the models, you can also calculate cost aggregation.
The metrics details also expose whether the requests have been cached by {{site.base_gateway}}, saving the cost of contacting the LLM providers, which improves performance.

Kong AI Gateway exposes metrics related to Kong and proxied upstream services in 
[Prometheus](https://prometheus.io/docs/introduction/overview/) 
exposition format, which can be scraped by a Prometheus Server.

The metrics are available on both the [Admin API](/api/gateway/admin-ee/) and the 

> [Status API](/api/gateway/status/)  at the `http://<host>:<port>/metrics` endpoint. 

Reviewer: ```suggestion
[Status API](/api/gateway/status/)  at the `http://{host}:{port}/metrics` endpoint. 
```

Note that the URL to those APIs is specific to your
installation. See [Accessing the metrics](#accessing-the-metrics) for more information.

The [Prometheus plugin](/plugins/prometheus/) records and exposes metrics at the node level. Your Prometheus
server will need to discover all Kong nodes via a service discovery mechanism,
and consume data from each node's configured `/metrics` endpoint.

AI metrics exported by the plugin can be graphed in Grafana using a [drop-in
dashboard](https://grafana.com/grafana/dashboards/21162-kong-cx-ai/).

> 

Reviewer: Sounds like this isn't available on konnect - we should probably indicate that with a note somehow.

Author: I assumed it worked in the same way on any data plane, but I'll double-check

Reviewer: Ohh I see where all this came from. It's mostly a word-for-word copy of the Prometheus plugin: https://docs.konghq.com/hub/kong-inc/prometheus/

And that's why it only talks about on-prem - that's what that doc was written for. 

I'm not 100% sure how the Konnect config works here, but I know that it does (eg see https://kongstrong.slack.com/archives/C06GB3KM155/p1736269876007219)

Author: So after looking around and testing a bit:
- I realized that I somehow didn't know that data planes can't use the Admin API
- The Status API iworks with the DP, I managed to make it work with the config I found [here](https://kongstrong.slack.com/archives/C04UMEJRZRP/p1680268002648999), so I'll add a note that the Status API is the only method compatible with Konnect

## Available metrics

When the `config.ai_metrics` parameter is set to `true` in the Prometheus plugin, you can get the following AI LLM metrics:

> {% if_version lte: 3.7.x %}

Reviewer: This doesn't seem to be working: https://deploy-preview-592--kongdeveloper.netlify.app/ai-gateway/monitor-ai-llm-metrics/3.7/

- **AI Cost**: AI costs charged by LLM providers.
  These are available per provider, model, cache, database name (if cached), and Workspace.
- **AI Tokens**: AI tokens counted by LLM providers.
  These are available per provider, model, cache, database name (if cached), token type, and Workspace.
{% endif_version %}

{% if_version gte: 3.8.x %}
- **AI Requests**: AI request sent to LLM providers.
- **AI Cost**: AI cost charged by LLM providers.
- **AI Tokens**: AI tokens counted by LLM providers.
- **AI LLM Latency**: Time taken to return a response by LLM providers.
- **AI Cache Fetch Latency**: Time taken to return a response from the cache.
- **AI Cache Embeddings Latency**: Time taken to generate embedding during the cache.

These metrics are available per provider, model, cache, database name (if cached), embeddings provider (if cached), embeddings model (if cached), and Workspace. The AI Tokens metrics are also available per token type.
{% endif_version %}

AI metrics are disabled by default as it may create high cardinality of metrics and may
cause performance issues. To enable them:

> * Set `config.ai_metrics` to `true` in the Prometheus plugin configuration.
> * Set `config.logging.log_statistics` to `true` in the AI Proxy or AI Proxy Advanced plugin.

Reviewer: ```suggestion
* Set `config.ai_metrics` to `true` in the [Prometheus plugin configuration](/plugins/prometheus/reference/).
* Set `config.logging.log_statistics` to `true` in the [AI Proxy](/plugins/ai-proxy/reference/) or [AI Proxy Advanced plugin](/plugins/ai-proxy-advanced/reference/).
```

# HELP ai_cache_embeddings_latency AI cache latencies per ai_provider/database in Kong
# TYPE ai_cache_embeddings_latency bucket
ai_cache_embeddings_latency{ai_provider="provider1",ai_model="model1",cache_status="hit",vector_db="redis",embeddings_provider="openai",embeddings_model="text-embedding-3-large",Workspace="workspace1",le="+Inf"} 2
# HELP ai_llm_provider_latency AI cache latencies per ai_provider/database in Kong
# TYPE ai_llm_provider_latency bucket
ai_llm_provider_latency{ai_provider="provider1",ai_model="model1",cache_status="hit",vector_db="redis",embeddings_provider="openai",embeddings_model="text-embedding-3-large",Workspace="workspace1",le="+Inf"} 2
```

{:.note}
> **Note:** If you don't use any cache plugins, then `cache_status`, `vector_db`,
`embeddings_provider`, and `embeddings_model` values will be empty. 

## Accessing the metrics

In most configurations, the Kong Admin API will be behind a firewall or would
need to be set up to require authentication. Here are a couple of options to
allow access to the `/metrics` endpoint to Prometheus:

> * If the Status API is enabled with the `status_listen` parameter in the [{{site.base_gateway}} configuration](/gateway/configuration/), 

Reviewer: ```suggestion
* If the Status API is enabled with the `status_listen` parameter in the [{{site.base_gateway}} configuration](/gateway/configuration/#status_listen), 
```

then its `/metrics` endpoint can be used. This is the preferred method.

* The `/metrics` endpoint is also available on the Admin API, which can be used

> if the Status API is not enabled. Note that this endpoint is unavailable
> when [RBAC](/api/gateway/admin-ee/3.9/#/operations/get-rbac-users) is enabled on the
> Admin API (Prometheus does not support key authentication to pass the token).

Reviewer: ```suggestion
if the Status API is not enabled. Note that this endpoint is unavailable
when [RBAC](/api/gateway/admin-ee/#/operations/get-rbac-users) is enabled on the
Admin API, as Prometheus doesn't support key authentication to pass the RBAC token.
```
